docs/tools/testing/unit_testing.md

```diff
@@ -1,27 +1,275 @@
+}
+uint32_t osSemaphoreGetCount(osSemaphoreId_t semaphore_id)
+{
+    return count;
+}
+osSemaphoreId_t osSemaphoreNew(uint32_t max_count, uint32_t initial_count, const osSemaphoreAttr_t *attr)
+{
+    return (void *)&count; // Just a dymmy reference
+}
+
+class  TestSemaphore : public  testing::Test {
+protected:
+    rtos::Semaphore *sem;
+
+    virtual  void  SetUp()
+    {
+        sem = new rtos::Semaphore();
+    }
+
+    virtual  void  TearDown()
+    {
+        delete sem;
+    }
+};
+
+TEST_F(TestSemaphore, constructor)
+{
+    EXPECT_TRUE(sem);
+}
+```
+
+### Building and running unit tests
+
+#### Testing with Mbed CLI
+
+You can build and run unit tests through Arm Mbed CLI. Mbed CLI supports unit testing through `mbed test --unittests` command with the same arguments. You can also use the tool to generate new test files. To learn how to do so, please see [a handbook section about using unit tests with Mbed CLI](https://os.mbed.com/docs/latest/tools/test-and-debug.html#unit-testing). For other information on using Mbed CLI, please see the [CLI documentation](https://os.mbed.com/docs/latest/tools/arm-mbed-cli.html).
+
+#### Build manually without Python tools
+
+1. Create a build directory e.g. `mkdir UNITTESTS/build`.
+2. Move to the build directory `cd UNITTESTS/build`.
+3. Run CMake with `cmake [RELATIVE PATH TO UNITTESTS DIR] [OPTIONAL ARGUMENTS]` e.g. `cmake ..`:
+	* Add `-g [generator]` argument if target other than Unix Makefiles e.g. MinGW `-g "MinGW Makefiles"`
+4. Run make program (make, gmake, mingw32-make, etc).
+
+##### Run CMake and build Unix Makefiles (GNU make)
+
+```
+cmake ..
+make
+```
+
+##### Run CMake and build MinGW Makefiles (mingw32-make)
+
+```
+cmake -G "MinGW Makefiles" ..
+mingw32-make
+```
+
+##### Custom CMake variables
+
+Usage: 
+`cmake [RELATIVE PATH TO UNITTESTS DIR] [OPTIONS]`
+
+Keyword variables (usage `cmake -D<VARIABLE>(:<TYPE>)=<value>`:
+
+| Variable | Type | Accepted values | Description |
+| -------- | ---- | --------------- | ----------- |
+| COVERAGE | STRING | merged<br>separate | Generate merged or individual reports |
+
+#### Run in terminal
+
+Unit tests can be run separately from each executable or by using ctest test runner. Run ctest with make program using target test. Options can be passed to ctest using ARGS argument. See [ctest manual](https://cmake.org/cmake/help/v3.0/manual/ctest.1.html) for more information.
+
+Run ctest on test suite level:
+```
+{MAKE_PROGRAM} test -C [RELATIVE PATH TO BUILD DIRECTORY]
+```
+e.g. `make test -C UNITTESTS/build` or `mingw32-make test -C UNITTESTS/build`
+
+Run ctest verbose (show each test case):
+```
+{MAKE_PROGRAM} test -C UNITTESTS/build ARGS="-V"
+```
+
+Run ctest dashboard test and create test results:
+```
+{MAKE_PROGRAM} test --C UNITTESTS/build ARGS="-D ExperimentalTest"
+```
+
+#### Run with GUI test runner
+
+1. Build and/or install *gtest-runner* using the [documentation](https://github.com/nholthaus/gtest-runner).
```

Comment:
Build and install are never supposed to be in the same bullet. Tell me to install if I don't have it, then in the next bullet tell me to build.